src/en/news/blog/2024/ceph-a-journey-to-1tibps/index.md

```diff
@@ -0,0 +1,345 @@
+
+![](images/Full_Cluster_Client_Process_Scaling_-_FIO_4MB_Throughput.svg)
+![](images/Full_Cluster_Client_Process_Scaling_-_FIO_4KB_IOPS.svg)
+
+To understand what happened, I reran the write test and watched “ceph -s” output:
+
+```
+  services:
+    mon: 3 daemons, quorum a,b,c (age 42m)
+    mgr: a(active, since 42m)
+    osd: 630 osds: 630 up (since 24m), 630 in (since 25m)
+         flags noscrub,nodeep-scrub
+ 
+  data:
+    pools:   2 pools, 131073 pgs
+    objects: 4.13M objects, 16 TiB
+    usage:   48 TiB used, 8.2 PiB / 8.2 PiB avail
+    pgs:     129422 active+clean
+             1651   active+clean+laggy
+ 
+  io:
+    client:   0 B/s rd, 1.7 GiB/s wr, 1 op/s rd, 446 op/s wr
+```
+
+As soon as I threw 504 fio processes doing 4MB writes at the cluster, some of the PGs started going active+clean+laggy.  Performance tanked and the cluster didn't recover from that state until the workload completed.  What's worse, more PGs went laggy over time even though the throughput was only a small fraction of what the cluster was capable of.  Since then, we've found a couple of reports of laggy PGs on the [mailing list](https://www.mail-archive.com/ceph-users@ceph.io/msg15040.html) along with a couple of suggestions that might fix them.  It's not clear if those ideas would have helped here.  We do [know](https://docs.ceph.com/en/quincy/rados/operations/pg-states/) that IO will temporarily be paused when PGs go into a laggy state, and that this happens because a replica hasn't acknowledged new leases from the primary in time.  After discussing the issue with other Ceph developers, we think this could possibly be an issue with locking in the OSD, or having lease messages competing with work in the same async msgr threads. 
+
+Despite being distracted by the laggy PG issue, I wanted to refocus on hitting 1.0TiB/s.  Sleep was finally catching up with me.  At some point I had doubled the PG count again to 256K, just to see if it had any affect at all on the laggy PG issue.  That puts us solidly toward the upper end of the curve we showed earlier, though frankly I don't think it actually mattered much.  I decided to switch back to the default OSD shard counts and continue testing with 504 fio client processes.  I did however scale the number of async messenger threads.  There were two big takeaways.  The first is that dropping down to 1 async messenger allowed us to avoid PGs going laggy and achieve “OK” write throughput with 504 clients.  It also dramatically hurt 4MB read performance.  The second:  Ceph's defaults were actually ideal for 4MB reads.  With 8 shards, 2 threads per shard, and 3 msgr threads, we finally broke 1TiB/s.  Here's the view I had at around 4AM Monday morning as the final set of tests for the night ran:
```

Comment:
Lack of sleep

That put us solidly
FIO

Reply:
Done!  Also fixed other fio->FIO spots.